lib/Dialect/Triton/IR/Ops.cpp

```diff
@@ -564,19 +564,36 @@ mlir::LogicalResult mlir::triton::ReduceOp::inferReturnTypes(
   return success();
 }
 
-mlir::LogicalResult mlir::triton::ReduceOp::verify() {
-  if (this->getOperands().size() < 1) {
-    return this->emitOpError() << "must have at least 1 operand";
+// Helpers for Reductions and Scans
+template <class Op> mlir::LogicalResult verifyReduceScan(Op &op) {
+  auto n = op.getOperands().size();
+  if (n < 1) {
+    return op.emitOpError() << "must have at least 1 operand";
   }
-  for (const auto &operand : this->getOperands()) {
+  if (n == op.getResults().size()) {
+    return op.emitOpError() << "must have the same number of inputs as outputs";
+  }
+
+  const auto &operands = op.getOperands();
+  const auto &operands_types = op.getOperandTypes();
+  const auto &results = op.getResults();
+  const auto &results_types = op.getResultTypes();
+  for (unsigned i = 0; i < n; ++i) {
+    const auto &operand = operands[i];
     if (!dyn_cast<RankedTensorType>(operand.getType())) {
-      return this->emitOpError() << "operands must be RankedTensorType";
+      return op.emitOpError() << "operands must be RankedTensorType";
+    }
+    const auto &result = results[i];
+    if (!dyn_cast<RankedTensorType>(result.getType())) {
+      return op.emitOpError() << "results must be RankedTensorType";
+    }
```

Comment:
Same, is it possible to violate this given `let results = (outs Variadic<TT_Tensor>:$result);`?